diff -b -u -r libshout-2.4.1-a/src/shout.c libshout-2.4.1-b/src/shout.c
--- libshout-2.4.1-a/src/shout.c	2015-11-09 06:54:12.000000000 -0500
+++ libshout-2.4.1-b/src/shout.c	2016-11-25 23:08:02.000000000 -0500
@@ -130,6 +130,7 @@
 		return NULL;
 	}
         self->tls_mode = SHOUT_TLS_AUTO;
+        self->tls_mode_used = SHOUTERR_NOTLS;
 #endif
 
 	self->port = LIBSHOUT_DEFAULT_PORT;
@@ -378,7 +379,7 @@
 		return SHOUTERR_NOCONNECT;
 
 #ifdef HAVE_OPENSSL
-	switch (self->tls_mode) {
+	switch (self->tls_mode_used) {
 	case SHOUT_TLS_DISABLED:
 		/* nothing to do */
 	break;
@@ -1157,22 +1158,45 @@
 
 	case SHOUT_STATE_TLS_PENDING:
 #ifdef HAVE_OPENSSL
-		if (self->tls_mode == SHOUT_TLS_DISABLED) {
-			/* nothing to be done */
-		} else if (self->tls_mode == SHOUT_TLS_AUTO || self->tls_mode == SHOUT_TLS_AUTO_NO_PLAIN) {
+           if (self->tls_mode_used < 0) {
+               switch (self->tls_mode) {
+                   case SHOUT_TLS_DISABLED:
+                   case SHOUT_TLS_RFC2818:
+                   case SHOUT_TLS_RFC2817:
+                       self->tls_mode_used = self->tls_mode;
+                   break;
+                   case SHOUT_TLS_AUTO:
+                   case SHOUT_TLS_AUTO_NO_PLAIN:
 			if (self->server_caps & LIBSHOUT_CAP_GOTCAPS) {
-				/* We had a probe allready, otherwise just do nothing to poke the server. */
+                           /* We had a probe allready, otherwise just poke the server. */
 				if (self->server_caps & LIBSHOUT_CAP_UPGRADETLS) {
-					self->tls_mode = SHOUT_TLS_RFC2817;
+                               self->tls_mode_used = SHOUT_TLS_RFC2817;
 				} else {
-					if (self->tls_mode == SHOUT_TLS_AUTO_NO_PLAIN)
+                               if (self->tls_mode == SHOUT_TLS_AUTO_NO_PLAIN) {
+                                   self->tls_mode_used = SHOUTERR_NOTLS;
 						return SHOUTERR_NOTLS;
-					self->tls_mode = SHOUT_TLS_DISABLED;
+                               }
+                               self->tls_mode_used = SHOUT_TLS_DISABLED;
 				}
 				self->state = SHOUT_STATE_TLS_PENDING;
 				goto retry;
+                       } else {
+                           /* TODO: do something. */
+                               if (self->tls_mode == SHOUT_TLS_AUTO_NO_PLAIN) {
+                                       self->tls_mode_used = SHOUTERR_NOTLS;
+                                       return SHOUTERR_NOTLS;
+                               }
+                               self->tls_mode_used = SHOUT_TLS_DISABLED;
 			}
-		} else if (self->tls_mode == SHOUT_TLS_RFC2818 || self->upgrade_to_tls) {
+                   break;
+                   default:
+                       rc = SHOUTERR_INSANE;
+                       goto failure;
+                   break;
+
+			}
+           }
+           if (self->tls_mode_used == SHOUT_TLS_RFC2818 || self->upgrade_to_tls) {
 			if (!self->tls) {
 				self->tls = shout_tls_new(self, self->socket);
 				if (!self->tls) /* just guessing that it's a malloc error */
@@ -1183,7 +1207,7 @@
 					return SHOUTERR_BUSY;
 				goto failure;
 			}
-		} else if (self->tls_mode == SHOUT_TLS_RFC2817) {
+		} else if (self->tls_mode_used == SHOUT_TLS_RFC2817) {
 			if ((rc = shout_create_http_request_upgrade(self, "TLS/1.0")) != SHOUTERR_SUCCESS) {
 				if (rc == SHOUTERR_BUSY)
 					return SHOUTERR_BUSY;
@@ -1191,9 +1215,6 @@
 			}
 			self->state = SHOUT_STATE_REQ_PENDING;
 			goto retry;
-		} else {
-                        rc = SHOUTERR_INSANE;
-                        goto failure;
 		}
 #endif
 		self->state = SHOUT_STATE_REQ_CREATION;
@@ -1233,7 +1254,7 @@
 		} else if (rc == SHOUTERR_SOCKET && !(self->server_caps & LIBSHOUT_CAP_GOTCAPS) &&
 			   (self->tls_mode == SHOUT_TLS_AUTO || self->tls_mode == SHOUT_TLS_AUTO_NO_PLAIN)) {
 			self->state = SHOUT_STATE_RECONNECT;
-			self->tls_mode = SHOUT_TLS_RFC2818;
+			self->tls_mode_used = SHOUT_TLS_RFC2818;
 			goto retry;
 #endif
 		}
diff -b -u -r libshout-2.4.1-a/src/shout_private.h libshout-2.4.1-b/src/shout_private.h
--- libshout-2.4.1-a/src/shout_private.h	2015-11-09 06:54:12.000000000 -0500
+++ libshout-2.4.1-b/src/shout_private.h	2016-11-25 23:08:09.000000000 -0500
@@ -130,6 +130,7 @@
 #ifdef HAVE_OPENSSL
 	int upgrade_to_tls;
         int tls_mode;
+        int tls_mode_used;
         char *ca_directory;
         char *ca_file;
         char *allowed_ciphers;
